Conversation quickstart (#4493)
dapr/docs: daprdocs/content/en/getting-started/quickstarts/conversation-quickstart.md

---
type: docs
title: "Quickstart: Conversation"
linkTitle: Conversation
weight: 90
description: Get started with the Dapr conversation building block
---

{{% alert title="Alpha" color="warning" %}}
The conversation building block is currently in **alpha**. 
{{% /alert %}}

> Let's take a look at how the [Dapr conversation building block]({{< ref conversation-overview.md >}}) makes interacting with the LLM component easier. In this quickstart, you use the echo component to communicate with the LLM and ask it for a poem about Dapr. 

Reviewer: ```suggestion
Let's take a look at how the [Dapr conversation building block]({{< ref conversation-overview.md >}}) makes interacting with Large Language Models (LLMs) easier. In this quickstart, you use the echo component to communicate with the mock LLM and ask it a question about Dapr. 
```

You can try out this conversation quickstart by either:

- [Running the application in this sample with the Multi-App Run template file]({{< ref "#run-the-app-with-the-template-file" >}}), or
- [Running the application without the template]({{< ref "#run-the-app-without-the-template" >}})

> ## Run the app with the template file

Reviewer: We will also have python ready Hannah FYI

Author: @alicejgibbons do we have Python SDK ready? we typically document the SDK instructions and link to HTTP at the bottom of the doc as next steps.

Reviewer: Sorry we just have the python HTTP example then https://github.com/dapr/quickstarts/tree/release-1.15/conversation/python/http

Author: Ok! I'll make sure that's linked at the bottom of the doc!

{{< tabs ".NET" Go >}}

 <!-- .NET -->
{{% codetab %}}

### Step 1: Pre-requisites

For this example, you will need:

- [Dapr CLI and initialized environment](https://docs.dapr.io/getting-started).

> - [.NET SDK or .NET 6 SDK installed](https://dotnet.microsoft.com/download).

Reviewer: ```suggestion
- [.NET 8 SDK+ installed](https://dotnet.microsoft.com/download).
```

<!-- IGNORE_LINKS -->
- [Docker Desktop](https://www.docker.com/products/docker-desktop)
<!-- END_IGNORE -->

### Step 2: Set up the environment

Clone the [sample provided in the Quickstarts repo](https://github.com/dapr/quickstarts/tree/master/conversation).

```bash
git clone https://github.com/dapr/quickstarts.git
```

From the root of the Quickstarts directory, navigate into the conversation directory:

```bash
cd conversation/csharp/sdk
```

### Step 3: Launch the conversation service

Start the conversation service with the following command:

```bash
dapr run -f .
```

**Expected output**

```
== APP - conversation == Input sent: What is dapr?
== APP - conversation == Output response: What is dapr?
```

### What happened?

When you ran `dapr init` during Dapr install, the following YAML files were generated in the `.dapr/components` directory: 
- [`dapr.yaml` Multi-App Run template file]({{< ref "#dapryaml-multi-app-run-template-file" >}})

> - [`pubsub.yaml` component file]({{< ref "#pubsubyaml-component-file" >}})

Reviewer: pubsub.yaml? Isnt this conversation.yaml as per: https://github.com/dapr/quickstarts/tree/release-1.15/conversation/components

Author: yes, you're right - left over from copy/paste. Thanks for catching!

> - [`pubsub.yaml` component file]({{< ref "#pubsubyaml-component-file" >}})

Reviewer 2: I think this is correct. Check the reference link
```suggestion
- [`conversation.yaml` component file]({{< ref "#conversation-component-file" >}})
```

Running `dapr run -f .` in this Quickstart started the [conversation Program.cs]({{< ref "#programcs-conversation-app" >}}).

#### `dapr.yaml` Multi-App Run template file

Running the [Multi-App Run template file]({{< ref multi-app-dapr-run >}}) with `dapr run -f .` starts all applications in your project. This Quickstart has only one application, so the `dapr.yaml` file contains the following: 

```yml
version: 1
common:
  resourcesPath: ../../components/
apps:
  - appDirPath: ./conversation/
    appID: conversation
    daprHTTPPort: 3500
    command: ["dotnet", "run"]
```

> #### `conversation.yaml` LLM component

Reviewer: ```suggestion
#### Echo mock LLM component
```

In [`conversation/components`](https://github.com/dapr/quickstarts/tree/master/conversation/components), the [`conversation.yaml` file](https://github.com/dapr/quickstarts/tree/master/conversation/components/conversation.yml) configures the echo mock LLM component. 

```yml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: echo
spec:
  type: conversation.echo
  version: v1
```

> 

Reviewer: We need to add a paragraph here referencing how to use another component. For example:

#### Use the OpenAI component

To interface with a real LLM use one of the other [supported conversation components](https://v1-15.docs.dapr.io/reference/components-reference/supported-conversation/openai/) including OpenAI, Hugging Face, Anthropic, DeepSeek and more. For example, swap out the `echo` component with an `OpenAI` one, by replacing the `conversation.yaml` file with the following. You will need to copy your API key into the component file.

```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: openai
spec:
  type: conversation.openai
  metadata:
  - name: key
    value: <REPLACE_WITH_YOUR_KEY>
  - name: model
    value: gpt-4-turbo
  - name: cacheTTL
    value: 10m
```

Author: I like this, but I'd like to put it somewhere else - I don't think a quickstart doc is for adding context with other examples. I'll check the docs we have and see where it can fit.

Author: Adding to the how-to, since we use echo there too, and it's a better doc format for adding context. I can link to the section from the quickstart, as well.

```

{{% /codetab %}}

 <!-- Go -->
{{% codetab %}}

### Step 1: Pre-requisites

For this example, you will need:

- [Dapr CLI and initialized environment](https://docs.dapr.io/getting-started).
- [Latest version of Go](https://go.dev/dl/).
<!-- IGNORE_LINKS -->
- [Docker Desktop](https://www.docker.com/products/docker-desktop)
<!-- END_IGNORE -->

### Step 2: Set up the environment

Clone the [sample provided in the Quickstarts repo](https://github.com/dapr/quickstarts/tree/master/conversation).

```bash
git clone https://github.com/dapr/quickstarts.git
```

From the root of the Quickstarts directory, navigate into the conversation directory:

```bash
cd conversation/go/sdk
```

### Step 3: Launch the conversation service

Start the conversation service with the following command:

```bash
dapr run -f .
```

**Expected output**

```
== APP - conversation == Input sent: What is dapr?
== APP - conversation == Output response: What is dapr?
```

### What happened?

When you ran `dapr init` during Dapr install, the following YAML files were generated in the `.dapr/components` directory: 
- [`dapr.yaml` Multi-App Run template file]({{< ref "#dapryaml-multi-app-run-template-file" >}})

> - [`pubsub.yaml` component file]({{< ref "#pubsubyaml-component-file" >}})

Reviewer 2: Need to fix the component file reference here also

Running `dapr run -f .` in this Quickstart started [conversation.go]({{< ref "#programcs-conversation-app" >}}).

#### `dapr.yaml` Multi-App Run template file

Running the [Multi-App Run template file]({{< ref multi-app-dapr-run >}}) with `dapr run -f .` starts all applications in your project. This Quickstart has only one application, so the `dapr.yaml` file contains the following: 

```yml
version: 1
common:
  resourcesPath: ../../components/
apps:
  - appDirPath: ./conversation/
    appID: conversation
    daprHTTPPort: 3501
    command: ["go", "run", "."]
```

> #### `conversation.yaml` LLM component

Reviewer: same comments as above ^^

In [`conversation/components`](https://github.com/dapr/quickstarts/tree/master/conversation/components) directly of the quickstart, the [`conversation.yaml` file](https://github.com/dapr/quickstarts/tree/master/conversation/components/conversation.yml) configures the echo LLM component. 

```yml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: echo
spec:
  type: conversation.echo
  version: v1
```

> 

Reviewer 2: Also add the OpenAI example into here, as Alice showed above ^^

Author: I mentioned above, I like adding an example of another 'actual' component, but I don't think it belongs in a quickstart doc. I'm thinking the how-to, since that also uses an echo mock component and would be better suited for context like other examples.

#### `conversation.go` conversation app

In the application code:
- The app sends an input "What is dapr?" to the echo mock LLM component.
- The mock LLM echoes "What is dapr?". 

```go
package main

import (
	"context"
	"fmt"
	"log"

	dapr "github.com/dapr/go-sdk/client"
)

func main() {
	client, err := dapr.NewClient()
	if err != nil {
		panic(err)
	}

	input := dapr.ConversationInput{
		Message: "What is dapr?",
		// Role:     nil, // Optional
		// ScrubPII: nil, // Optional
	}

	fmt.Println("Input sent:", input.Message)

	var conversationComponent = "echo"

	request := dapr.NewConversationRequest(conversationComponent, []dapr.ConversationInput{input})

	resp, err := client.ConverseAlpha1(context.Background(), request)
	if err != nil {
		log.Fatalf("err: %v", err)
	}

	fmt.Println("Output response:", resp.Outputs[0].Result)
}
```

{{% /codetab %}}

{{< /tabs >}}

## Run the app without the template

{{< tabs ".NET" Go >}}

 <!-- .NET -->
{{% codetab %}}

### Step 1: Pre-requisites

For this example, you will need:

- [Dapr CLI and initialized environment](https://docs.dapr.io/getting-started).

> - [.NET SDK or .NET 6 SDK installed](https://dotnet.microsoft.com/download).

Reviewer: ```suggestion
- [.NET 8+ SDK installed](https://dotnet.microsoft.com/download).
```
